File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.dart
class Solution {
  double findMaxAverage(List<int> nums, int k) {
    int l = 0;
    double max_average = double.negativeInfinity;
    double sum = 0;
    int count = 0;

    for (int r = 0; r < nums.length; ++r) {
      // sum = 1
      // sum = 13
      // sum = 8
      // sum = 2

      // sum = 51
      sum += nums[r];
      // count = 1
      // count = 2
      // count = 3
      // count = 4

      // count = 4
      count++;

      while (count >= k) {
        // max_average = 2
        // max_average = 51
        max_average = max(max_average, sum);
        // count = 3
        count--;
        // sum = 1
        sum = sum - nums[l];
        // l = 1
        l += 1;
      }
    }
    return max_average / k;
  }
}

File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.py
class Solution:
    def findMaxAverage(self, nums: list[int], k: int) -> float:
        n = len(nums)
        cur_sum = 0        
        # this is simply build the initiate the window
        for i in range(k):
            cur_sum += nums[i]
        
        max_sum = cur_sum
        
        # after we calculate the init window 
        # we just have to move the start by one step => nums[i - k]
        # also the end by one step nums[i]
        for i in range(k, n):
            cur_sum = cur_sum + nums[i] - nums[i - k]
            max_sum = max(max_sum, cur_sum)
            
        return max_sum / k